File: tests/e2e/compose-approvals.js
const http = require("http");
const L = require("./lib");
const { assert, BASE, sleep, dismissUpdate } = L;

// Compose commands run by an assistant, and the approval a confirm action
// asks for: the assistant reaches POST /projects/:name/docker/compose over the
// local API socket the way its `compose-start` does (the socket is mounted
// into the runner, APISOCK names it, the X-Dev-Cockpit-Assistant header says
// who calls), a command marked to ask first parks the run instead of starting
// it (the answer carries `pending`, the run page reads Awaiting approval with
// its Cancel showing), the question is news of its own (`approval:<run>`,
// "Assistant asks approval.", read through the request API because an open
// page reads the entry by showing the dialog) and stands on any page as the
// app-wide dialog (@dc/gitprompt's second kind: the assistant, the project,
// the stack and the command, Approve and Deny, one "Approve and don't ask
// again" box). Deny ends the run declined and the assistant's thread holds a
// grey compose note saying so (data-assistant-compose="declined"); Approve
// runs it, the note says done, and the next confirm action still asks, and
// nothing about an assistant's run rings the bell (no docker:<project> entry
// for it). The Compose actions approval under /settings/assistant/approvals
// is one switch for every assistant (on by default, off lets every confirm
// action run, the socket refused, on the Docker settings too), the dialog's
// box turns it off for every assistant, and the trigger form hides any or all
// for a compose event.
//
// Fixture the host prepares before the run: the instance's projects dir holds
// a project named DOCKER_PROJECT (default "dockere2e") with a compose file of
// one service "web" on a local image, already up, the instance reaches the
// daemon and the docker CLI, has a coder installed (the fakes suffice) and
// its default compose actions (the confirm action is "Compose down with
// volumes", which the checks approve once, so the fixture goes down and is
// brought back up over the socket at the end). The runner needs the
// instance's local API socket mounted, e.g. -v <state-dir>/api:/apisock with
// APISOCK=/apisock/s, because nothing but the socket may act as an assistant.

const NAME = process.env.DOCKER_PROJECT || "dockere2e";
const APISOCK = process.env.APISOCK || "/apisock/s";
const DIALOG = ".swal2-popup:not(.swal2-toast)";

// asAssistant posts one form over the local socket as the assistant, which
// is exactly what the generated `cockpit` wrapper does for a turn.
function asAssistant(id, path, body) {
  return new Promise((resolve, reject) => {
    const data = new URLSearchParams(body).toString();
    const req = http.request({
      socketPath: APISOCK,
      path,
      method: "POST",
      headers: {
        "X-Dev-Cockpit-Assistant": id,
        "Content-Type": "application/x-www-form-urlencoded",
        "Content-Length": Buffer.byteLength(data),
        Accept: "application/json",
      },
    }, (res) => {
      let raw = "";
      res.on("data", (chunk) => { raw += chunk; });
      res.on("end", () => {
        let json = {};
        try { json = JSON.parse(raw); } catch {}
        resolve({ status: res.statusCode, json, raw });
      });
    });
    req.on("error", reject);
    req.end(data);
  });
}

async function csrfOf(page) {
  return page.evaluate(() => document.querySelector('meta[name="csrf-token"]')?.content || "");
}

async function runJSON(page, id) {
  const res = await page.request.get(`${BASE}/projects/${NAME}/docker/runs/${id}/output`, { headers: { Accept: "application/json" } });
  return res.json();
}

async function waitRun(page, id, done, what) {
  for (let i = 0; i < 120; i += 1) {
    const run = await runJSON(page, id);
    if (done(run)) return run;
    await sleep(500);
  }
  throw new Error(`run ${id} never ${what}`);
}

async function notifications(page) {
  const res = await page.request.get(`${BASE}/notifications`, { headers: { Accept: "application/json" } });
  return (await res.json()).notifications || [];
}

async function waitFor(fn, what, tries = 40) {
  for (let i = 0; i < tries; i += 1) {
    if (await fn()) return;
    await sleep(250);
  }
  throw new Error(`timed out waiting for ${what}`);
}

L.runFeature("COMPOSE APPROVALS", async ({ page, run }) => {
  let assistantID = "";
  let parked = "";

  await run("an assistant is made for the checks", async () => {
    await page.goto(`${BASE}/projects`, { waitUntil: "domcontentloaded" });
    await dismissUpdate(page);
    await page.waitForSelector(`#project-${NAME} [data-chip-kind="docker"]`, { timeout: 8000 });
    const created = await page.request.post(`${BASE}/assistants/new`, { form: { csrf_token: await csrfOf(page), form: "new", coder: "claude" }, headers: { Accept: "application/json" } });
    assistantID = (await created.json().catch(() => ({}))).id || "";
    assert(assistantID, `no assistant for the checks: ${created.status()}`);
    return assistantID;
  });

  await run("a confirm command started by the assistant parks the run, rings as an approval and asks on any page", async () => {
    // Off every page first: the entry has to exist before a page shows the
    // dialog and reads it.
    await page.goto("about:blank");
    const before = (await notifications(page)).map((n) => n.id);
    const answer = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(answer.status === 200, `the socket answered ${answer.status}: ${answer.raw}`);
    assert(answer.json.pending === true && answer.json.run, `the confirm action did not park: ${answer.raw}`);
    parked = answer.json.run;
    let fresh = [];
    await waitFor(async () => {
      fresh = (await notifications(page)).filter((n) => !before.includes(n.id));
      return fresh.length > 0;
    }, "the approval entry");
    const entry = fresh.find((n) => n.targetId === `approval:${parked}`);
    assert(entry, `no entry for the approval, only ${JSON.stringify(fresh.map((n) => n.targetId))}`);
    assert(entry.title === "Assistant asks approval.", `the entry's title reads "${entry.title}"`);
    assert(/Compose down with volumes in dockere2e/.test(entry.detail || entry.body || ""), `the entry's line reads "${entry.detail || entry.body}"`);
    assert(!entry.read, "the entry was read before anybody saw it");
    assert((entry.url || "").endsWith(`/docker/runs/${parked}`), `the entry leads to ${entry.url}`);

    await page.goto(`${BASE}/projects/${NAME}/docker/runs/${parked}`, { waitUntil: "domcontentloaded" });
    await page.waitForSelector(`${DIALOG} .swal2-title`, { timeout: 15000 });
    const title = await page.textContent(`${DIALOG} .swal2-title`);
    assert(/asks for approval/.test(title), `the dialog's title reads "${title}"`);
    const details = await page.textContent(`${DIALOG} [data-approval-details]`);
    for (const word of ["Assistant", NAME, "Compose down with volumes"]) {
      assert(details.includes(word), `the dialog does not say ${word}: ${details}`);
    }
    const command = await page.textContent(`${DIALOG} [data-gitprompt-command]`);
    assert(command.includes("docker compose down -v"), `the dialog does not show the command line: ${command}`);
    assert(await page.locator(`${DIALOG} .swal2-confirm`).textContent() === "Approve", "the confirm button is not Approve");
    assert(await page.locator(`${DIALOG} .swal2-deny`).textContent() === "Deny", "the deny button is not Deny");
    const remember = (await page.locator(`${DIALOG} .swal2-checkbox`).textContent()).trim();
    assert(remember === "Approve and don't ask again", `the box reads "${remember}"`);
    assert(await page.locator(`${DIALOG} .swal2-radio`).isHidden(), "the dialog still offers scopes");
    assert(!(await page.locator(`${DIALOG} .swal2-checkbox input`).isChecked()), "the dialog does not start on asking again");
    assert(await page.locator(`${DIALOG} dl[data-approval-details] dt`).count() === 4, "the details are no list of four");
    // The page under it is the run's, already saying what it waits for.
    const status = await page.textContent("[data-run-status]");
    assert(status.trim() === "Awaiting approval", `the run page reads "${status}"`);
    assert(await page.locator("[data-run-stop]").isVisible(), "a parked run hides its Cancel");
    await waitFor(async () => (await notifications(page)).filter((n) => n.targetId === `approval:${parked}` && !n.read).length === 0, "the shown entry to read itself");
    return `run ${parked} parked`;
  });

  await run("Deny ends the run declined and tells the assistant in its thread", async () => {
    await page.click(`${DIALOG} .swal2-deny`);
    await page.waitForSelector(DIALOG, { state: "detached", timeout: 8000 });
    const run = await waitRun(page, parked, (r) => !r.pending, "left its question");
    assert(run.declined === true && run.failure === "declined by the user", `the denied run reads ${JSON.stringify(run)}`);
    await waitFor(async () => (await page.textContent("[data-run-status]")).trim() === "Declined by the user", "the page to say declined");
    await page.goto(`${BASE}/assistants/${assistantID}`, { waitUntil: "domcontentloaded" });
    await page.waitForSelector('[data-assistant-note="compose"][data-assistant-compose="declined"]', { timeout: 15000 });
    const headline = await page.textContent('[data-assistant-compose="declined"] [data-assistant-note-headline]');
    assert(headline.trim() === `Compose declined: Compose down with volumes on ${NAME}`, `the note reads "${headline}"`);
    const bell = (await notifications(page)).filter((n) => n.targetId === `docker:${NAME}` && !n.read);
    assert(bell.length === 0, "an assistant's run rang the project's docker target");
    return headline.trim();
  });

  await run("Approve runs it, the note says done, and the next confirm command still asks", async () => {
    const answer = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(answer.json.pending === true, `the second confirm action did not park: ${answer.raw}`);
    const id = answer.json.run;
    await page.waitForSelector(`${DIALOG} .swal2-confirm`, { timeout: 15000 });
    await page.click(`${DIALOG} .swal2-confirm`);
    await page.waitForSelector(DIALOG, { state: "detached", timeout: 8000 });
    const done = await waitRun(page, id, (r) => !r.pending && !r.running, "finished after the approval");
    assert(done.exited === true && done.exit === 0 && done.status === "Exit status 0", `the approved run reads ${JSON.stringify(done)}`);
    await page.waitForSelector('[data-assistant-note="compose"][data-assistant-compose="done"]', { timeout: 15000 });
    const headline = await page.textContent('[data-assistant-compose="done"] [data-assistant-note-headline]');
    assert(headline.trim() === `Compose done: Compose down with volumes on ${NAME}`, `the done note reads "${headline}"`);
    assert(await page.locator('[data-assistant-compose="done"]').locator("xpath=..").locator("a[href$='/docker/runs/" + id + "']").count() === 1, "the note carries no link to the run");

    const again = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(again.json.pending === true, `an approval without the box stopped the asking: ${again.raw}`);
    await page.waitForSelector(`${DIALOG} .swal2-deny`, { timeout: 15000 });
    await page.click(`${DIALOG} .swal2-deny`);
    await page.waitForSelector(DIALOG, { state: "detached", timeout: 8000 });
    await waitRun(page, again.json.run, (r) => !r.pending, "was declined");
    const up = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "up" });
    assert(up.status === 200 && up.json.run, `bringing the fixture back answered ${up.raw}`);
    await waitRun(page, up.json.run, (r) => !r.running, "brought the fixture back");
    assert((await notifications(page)).filter((n) => n.targetId === `docker:${NAME}` && !n.read).length === 0, "an assistant's run rang the project's docker target");
    return "approved once, asked again, fixture back up";
  });

  const composeSwitch = '[data-assistant-approval="compose-actions"] input[type="checkbox"]';
  const saveApprovals = () => Promise.all([
    page.waitForNavigation({ waitUntil: "domcontentloaded" }),
    page.click('#settings-assistant-approvals button[type="submit"]'),
  ]);

  await run("the Compose actions approval is one switch for every assistant, and the socket cannot move it or the compose commands", async () => {
    await page.goto(`${BASE}/settings/assistant/approvals`, { waitUntil: "domcontentloaded" });
    await dismissUpdate(page);
    assert(await page.locator("[data-assistant-approval]").count() === 1, "the tab lists more than the one approval");
    const label = (await page.textContent('[data-assistant-approval="compose-actions"]')).trim();
    assert(label.startsWith("Compose actions approval"), `the row reads "${label}"`);
    assert(await page.locator(`[data-assistant-approval="${assistantID}"]`).count() === 0, "the tab still carries a row per assistant");
    assert(await page.locator(composeSwitch).isChecked(), "the approval is not on by default");
    await page.locator(composeSwitch).uncheck();
    await saveApprovals();
    assert(!(await page.locator(composeSwitch).isChecked()), "the switch did not stay off");
    const free = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(free.status === 200 && free.json.pending !== true, `a confirm action was parked with the approval off: ${free.raw}`);
    await waitRun(page, free.json.run, (r) => !r.running, "finished");
    const refused = await asAssistant(assistantID, "/settings/assistant/approvals", { "approval-compose-actions": "1" });
    assert(refused.status === 403, `the socket moved the approval: ${refused.status}`);
    for (const path of ["/settings/docker", "/docker/actions/restore"]) {
      const docker = await asAssistant(assistantID, path, { docker_host: "" });
      assert(docker.status === 403, `the socket reached the compose commands on ${path}: ${docker.status} ${docker.raw}`);
    }
    await page.reload({ waitUntil: "domcontentloaded" });
    assert(!(await page.locator(composeSwitch).isChecked()), "the refused socket call moved the switch");
    await page.locator(composeSwitch).check();
    await saveApprovals();
    assert(await page.locator(composeSwitch).isChecked(), "the switch did not come back on");
    const up = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "up" });
    await waitRun(page, up.json.run, (r) => !r.running, "brought the fixture back");
    return "off, socket refused on the approval and the docker settings, back on";
  });

  await run("Approve and don't ask again turns the approval off for every assistant", async () => {
    const created = await page.request.post(`${BASE}/assistants/new`, { form: { csrf_token: await csrfOf(page), form: "new", coder: "claude" }, headers: { Accept: "application/json" } });
    const otherID = (await created.json().catch(() => ({}))).id || "";
    assert(otherID, `no second assistant: ${created.status()}`);
    const parkedRun = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(parkedRun.json.pending === true, `the confirm action did not park: ${parkedRun.raw}`);
    await page.waitForSelector(`${DIALOG} .swal2-checkbox input`, { timeout: 15000 });
    await page.check(`${DIALOG} .swal2-checkbox input`);
    await page.click(`${DIALOG} .swal2-confirm`);
    await page.waitForSelector(DIALOG, { state: "detached", timeout: 8000 });
    await waitRun(page, parkedRun.json.run, (r) => !r.pending && !r.running, "finished after the approval");
    const other = await asAssistant(otherID, `/projects/${NAME}/docker/compose`, { stack: "", action: "down-volumes" });
    assert(other.status === 200 && other.json.pending !== true, `another assistant was still asked: ${other.raw}`);
    await waitRun(page, other.json.run, (r) => !r.running, "finished");
    await page.goto(`${BASE}/settings/assistant/approvals`, { waitUntil: "domcontentloaded" });
    assert(!(await page.locator(composeSwitch).isChecked()), "the box left the approval on");
    await page.locator(composeSwitch).check();
    await saveApprovals();
    const up = await asAssistant(assistantID, `/projects/${NAME}/docker/compose`, { stack: "", action: "up" });
    await waitRun(page, up.json.run, (r) => !r.running, "brought the fixture back");
    const gone = await page.request.post(`${BASE}/assistants/${otherID}`, { form: { csrf_token: await csrfOf(page), form: "delete" }, headers: { Accept: "application/json" } });
    assert(gone.status() === 200, `the second assistant's delete answered ${gone.status()}`);
    return "off from the dialog for both, back on";
  });

  await run("the trigger form offers no any or all for a compose event", async () => {
    await page.goto(`${BASE}/assistants/triggers?form=new&assistant=${assistantID}`, { waitUntil: "domcontentloaded" });
    await page.waitForSelector("[data-trigger-event]", { timeout: 8000 });
    await page.selectOption("[data-trigger-event]", "compose-ended");
    const mode = page.locator("[data-trigger-mode]");
    assert(!(await mode.isVisible()), "a compose event shows the mode");
    assert(await page.locator("[data-trigger-mode] select").isDisabled(), "the hidden mode still posts");
    await page.selectOption("[data-trigger-event]", "job-closed");
    assert(await mode.isVisible(), "a job event hides the mode");
    return "mode only for job and coder";
  });

  await run("the assistant is taken away again", async () => {
    await page.goto(`${BASE}/projects`, { waitUntil: "domcontentloaded" });
    const gone = await page.request.post(`${BASE}/assistants/${assistantID}`, { form: { csrf_token: await csrfOf(page), form: "delete" }, headers: { Accept: "application/json" } });
    assert(gone.status() === 200, `the delete answered ${gone.status()}`);
  });
});

File: internal/web/composeapprovals.go
package web

import (
	"errors"
	"log"
	"path/filepath"
	"strings"
	"time"

	"github.com/marein/dev-cockpit/internal/askpass"
	"github.com/marein/dev-cockpit/internal/assistant"
	"github.com/marein/dev-cockpit/internal/coder"
	"github.com/marein/dev-cockpit/internal/docker"
	"github.com/marein/dev-cockpit/internal/eventbus"
	"github.com/marein/dev-cockpit/internal/notify"
	"github.com/marein/dev-cockpit/internal/project"
)

// An assistant runs the compose commands a person runs, over the same route,
// and a command that asks first asks the person first there too: the run is
// parked, the command answers the run id at once, and the question stands in
// the askpass broker under the run's own key, where every signed-in page
// shows it and the push channels carry it to a phone. The person approves or
// denies; no answer within the bound, or a restart that takes the question
// with it, ends the run declined. The assistant is told either way, in its
// own thread, see composeDone.

// composeApprovalTimeout bounds how long a parked run waits for the decision.
// Long enough for a phone in a pocket, short enough that a "down with
// volumes" approved much later still means what the assistant asked for.
const composeApprovalTimeout = 30 * time.Minute

// assistantAsksForCompose says whether a confirm action an assistant starts
// waits for the user: yes unless the user turned the Compose actions approval
// off, in the modal or in the Approvals settings. It is one switch for every
// assistant, so the owner decides nothing.
func (s *Server) assistantAsksForCompose() bool {
	return s.assistants != nil && ApprovalAsks(s.settings, approvalComposeActions)
}

// assistantLabel is what a question calls an assistant: its title cut to a
// label, or the surface's own word for one nobody named.
func (s *Server) assistantLabel(id string) string {
	if s.assistants == nil {
		return assistant.Name
	}
	entry, err := s.assistants.Get(id)
	if err != nil {
		return assistant.Name
	}
	if title := strings.TrimSpace(entry.Title); title != "" && title != assistant.DefaultTitle {
		return coder.ShortTitle(title)
	}
	return assistant.Name
}

// askComposeApproval registers the bridge of one parked run before the
// request that parked it is answered, so a cancel that comes after finds the
// bridge and takes it down. The question itself stands once
// awaitComposeApproval parks it on that bridge. A run that stopped being parked meanwhile, a cancel from the
// page in between, gets no question at all. A question that cannot be put up
// declines the run without a note, the error is the caller's whole answer.
func (s *Server) askComposeApproval(owner string, p project.Project, action docker.Action, run docker.RunView) (*askpass.Action, error) {
	var bridge *askpass.Action
	if s.askpassBroker != nil {
		bridge = s.askpassBroker.BeginApproval(askpass.ApprovalKey(run.ID), askpass.Question{
			Project:   p.Name,
			Action:    action.Label,
			Command:   run.Command,
			Dir:       run.Dir,
			Assistant: s.assistantLabel(owner),
			Stack:     stackLabel(p.Path, run.Dir),
			URL:       dockerRunPath(p.Name, run.ID),
		})
	}
	if bridge == nil {
		const reason = "the approval could not be asked"
		if err := s.docker.WithdrawCompose(run.ID, reason); err != nil {
			log.Printf("docker: withdraw run %s: %v", run.ID, err)
		}
		return nil, errors.New(reason)
	}
	if current, ok := s.docker.ComposeRunByID(run.ID); !ok || !current.Pending {
		bridge.End()
		return nil, nil
	}
	return bridge, nil
}

// declineParked ends a parked run, quietly where it is over already: a run
// called off while it waited had its end reported by the cancel.
func (s *Server) declineParked(id, reason string, byUser bool) {
	if current, ok := s.docker.ComposeRunByID(id); ok && !current.Pending {
		return
	}
	if err := s.docker.DeclineCompose(id, reason, byUser); err != nil {
		log.Printf("docker: decline run %s: %v", id, err)
	}
}

// awaitComposeApproval waits out the question askComposeApproval put up and
// starts or ends the run on what comes back. It runs off the request, like
// the run itself would: the command that asked is long answered when the
// person decides. Which of a decision and a cancel wins is the docker
// service's to settle, see ApproveCompose.
func (s *Server) awaitComposeApproval(bridge *askpass.Action, runID string) {
	// The end of the action is what a timeout looks like: the question
	// leaves the list and the asker reads no decision.
	expiry := time.AfterFunc(composeApprovalTimeout, bridge.End)
	decision, decided := bridge.AskApproval()
	expiry.Stop()
	bridge.End()
	switch {
	case decided && decision.Approved:
		// Don't ask again is about the kind, not about this run: it holds
		// even where a cancel took the run away while the box was ticked.
		if decision.Remember {
			setApprovalAsks(s.settings, approvalComposeActions, false)
		}
		if current, ok := s.docker.ComposeRunByID(runID); ok && !current.Pending {
			break
		}
		if err := s.docker.ApproveCompose(runID); err != nil {
			log.Printf("docker: approve run %s: %v", runID, err)
		}
	case decided:
		// Only the user answers an approval, the local socket is refused on
		// it, so a Deny is the user's own click and rings nobody.
		s.declineParked(runID, "declined by the user", true)
	default:
		s.declineParked(runID, "the approval expired unanswered", false)
	}
	s.bus.Publish(eventbus.Event{Type: "docker"})
}

// declineParkedRuns ends the parked runs match picks and takes their
// questions down with them, for an owner or a directory that is going away.
func (s *Server) declineParkedRuns(match func(docker.RunView) bool, reason string) {
	ids := s.docker.DeclinePending(match, reason)
	if len(ids) == 0 {
		return
	}
	if s.askpassBroker != nil {
		for _, id := range ids {
			if bridge := s.askpassBroker.Find(askpass.ApprovalKey(id)); bridge != nil {
				bridge.End()
			}
		}
	}
	s.bus.Publish(eventbus.Event{Type: "docker"})
}

// declineOwnerApprovals is what a deleted assistant leaves: nobody is there
// any more to hear how a run it asked for went.
func (s *Server) declineOwnerApprovals(owner string) {
	s.declineParkedRuns(func(run docker.RunView) bool { return run.Owner == owner }, "the assistant was deleted")
}

// declineProjectApprovals is what a deleted project leaves: a run approved
// later would start in a directory that is gone.
func (s *Server) declineProjectApprovals(root string) {
	root = filepath.Clean(root)
	s.declineParkedRuns(func(run docker.RunView) bool {
		return run.Dir == root || strings.HasPrefix(run.Dir, root+string(filepath.Separator))
	}, "the project was deleted")
}

// questionTarget is the notification entry a standing question holds while it
// stands, empty for one that holds none: a proxied git question under its
// project, an approval under its run. It is decided here and handed to the
// dialog with the question, never worked out again in the browser.
func questionTarget(q askpass.Question) string {
	if run, ok := askpass.ApprovalRun(q.Key); ok && q.Kind == askpass.KindApproval {
		return notify.ApprovalTarget(run)
	}
	if q.External {
		return notify.GitPromptTarget(q.Project)
	}
	return ""
}

// composeReport is what the assistant's thread is told about an owned run
// that ended: the run, where it ran, how it went, the tail of what it wrote
// and the page that shows the rest. The stack is named the way the compose
// menu names it; a project that is gone by then leaves the directory.
func (s *Server) composeReport(run docker.ComposeRun, err error, output string) assistant.ComposeReport {
	report := assistant.ComposeReport{
		Owner:    run.Owner,
		Run:      run.ID,
		Project:  run.Label,
		Stack:    run.Dir,
		Action:   run.Action,
		URL:      dockerRunPath(run.Label, run.ID),
		Failed:   run.Failed,
		Declined: run.Declined,
		Exited:   run.Exited,
		Exit:     run.Exit,
		Output:   output,
		ByUser:   run.ByUser,
	}
	if err != nil {
		report.Reason = err.Error()
	}
	if p, perr := s.projects.FindByName(run.Label); perr == nil {
		report.Stack = stackLabel(p.Path, run.Dir)
	}
	return report
}
